### PR TITLE
Add mgrctl commands "ssh" and "ssh knownhost remove"

### LINKED ISSUE #501: Command to cleanup SSH known hosts
Today, running tests, I ended up with a VM using the same IP from an old VM, but of course a different SSH key.

That caused the following while onboarding:

imagen

Maybe this should be an `mgrctl` command so the user doesn't need to manually run things on the container.

Something like:

`mgradm ssh remove-know-host <IP/HOSTNAME>`

### PR DESCRIPTION
**Pull request by marv7000**
## What does this PR change?

Adds a command to remove an entry from the list of Salt's known SSH hosts on a Uyuni server. Fixes #501 

## Codespace

Check if you already have a running container clicking on [Running CodeSpace](https://github.com/codespaces)

[Create CodeSpace](https://codespaces.new/uyuni-project/uyuni-tools)  [About billing for Github Codespaces](https://docs.github.com/en/billing/managing-billing-for-github-codespaces/about-billing-for-github-codespaces) [CodeSpace Billing Summary](https://github.com/settings/billing/summary) [CodeSpace Limit](https://github.com/settings/billing/spending_limit)

## Test coverage
- No tests: Interacts with server API.

- [x] **DONE**

## Links

Issue(s): #501 (and PR https://github.com/uyuni-project/uyuni/pull/11720)

- [x] **DONE**

## Changelogs

Make sure the changelogs entries you are adding are compliant with https://github.com/uyuni-project/uyuni/wiki/Contributing#changelogs and https://github.com/uyuni-project/uyuni/wiki/Contributing#uyuni-projectuyuni-repository

If you don't need a changelog check, please mark this checkbox:

- [ ] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)

# Before you merge

Check [How to branch and merge properly](https://github.com/uyuni-project/uyuni/wiki/How-to-branch-and-merge-properly)!

### CODE REVIEW

File: mgrctl/cmd/ssh/remove_known_host.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+func runRemoveKnownHost(_ *types.GlobalFlags, flags *apiFlags, _ *cobra.Command, args []string) error {
+	hostname := args[0]
+	// Default to port 22 if none is provided
+	port := "22"
+	if len(args) == 2 {
+		port = args[1]
+	}
+
+	log.Debug().Msgf("Removing host %s:%s", hostname, port)
+	client, err := api.Init(&flags.ConnectionDetails)
```
Comment (cbosdo) [inline mgrctl/cmd/ssh/remove_known_host.go]:
```suggestion
	client, err := apiInit(&flags.ConnectionDetails)
```

File: mgrctl/cmd/ssh/knownhost/remove.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+		err = client.Login()
+	}
+	if err != nil {
+		return utils.Errorf(err, L("unable to login to the server"))
+	}
+
+	path := "admin/ssh/removeKnownHost?hostname=" + hostname + "&port=" + port
+	var data map[string]interface{}
+
+	res, err := api.Post[interface{}](client, path, data)
```
Comment (cbosdo) [inline mgrctl/cmd/ssh/knownhost/remove.go:27]:
```suggestion
	res, err := apiPost[interface{}](client, path, data)
```

File: mgrctl/cmd/ssh/knownhost/remove.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+import (
+	"fmt"
+
+	"github.com/rs/zerolog/log"
+	"github.com/spf13/cobra"
+	"github.com/uyuni-project/uyuni-tools/shared/api"
+	. "github.com/uyuni-project/uyuni-tools/shared/l10n"
+	"github.com/uyuni-project/uyuni-tools/shared/types"
+	"github.com/uyuni-project/uyuni-tools/shared/utils"
+)
```
Comment (cbosdo) [inline mgrctl/cmd/ssh/knownhost/remove.go:17]:
```suggestion
)

var apiInit = api.Init
var apiPost = api.Post
```

This way you can set `apiInit` and apiPost to fake functions in unit tests.

File: mgrctl/cmd/ssh/remove_known_host.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+	if err == nil {
+		err = client.Login()
+	}
+	if err != nil {
+		return utils.Errorf(err, L("unable to login to the server"))
+	}
+
+	path := "admin/ssh/removeKnownHost?hostname=" + hostname + "&port=" + port
+	var data map[string]interface{}
```
[the remark this answers is not in the document]
Comment (cbosdo) [reply mgrctl/cmd/ssh/remove_known_host.go]: Indeed, the hostname and port need to be sanitized as this could lead to code injection

File: mgrctl/cmd/ssh/remove_known_host.go [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+		return utils.Errorf(err, L("error in query '%s'"), path)
+	}
+
+	if !res.Success {
+		log.Error().Msg(res.Message)
+	}
+
+	success, ok := res.Result.(float64)
+	if !ok {
+		return fmt.Errorf(L("unexpected server response '%s'"), string(res.Result.([]byte)))
```
[the remark this answers is not in the document]
Comment (cbosdo) [reply mgrctl/cmd/ssh/remove_known_host.go]: `%v` is probably good enough here, no need for %T on top of it.

File: mgrctl/cmd/ssh/ssh.go [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+func NewCommand(globalFlags *types.GlobalFlags) *cobra.Command {
+	var flags apiFlags
+
+	sshCmd := &cobra.Command{
+		Use:   "ssh",
+		Short: L("SSH management commands"),
+	}
+
+	sshRemoveKnownHostCmd := &cobra.Command{
+		Use:   "remove_known_host hostname [port]",
```
[the remark this answers is not in the document]
Comment (cbosdo) [reply mgrctl/cmd/ssh/ssh.go]: May be `knownhost remove` would be more consistent and opened to other operations on the knownhosts file if needed.

File: mgrctl/cmd/ssh/remove_known_host_test.go [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+	if err != nil {
+		t.FailNow()
+	}
+
+	client.Client = &mocks.MockClient{
+		DoFunc: MockKnownHostDo,
+	}
+
+	testutils.AssertEquals(t, "", nil, removeKnownHost(client, "client.uyuni.lan", "22"))
+	testutils.AssertEquals(t, "", nil, removeKnownHost(client, "escaped?client.uyuni.lan", "42"))
```
Comment (cbosdo) [inline mgrctl/cmd/ssh/remove_known_host_test.go]: Add message to understand what the failure is.
```suggestion
	testutils.AssertEquals(t, "unexpected error", nil, removeKnownHost(client, "client.uyuni.lan", "22"))
	testutils.AssertEquals(t, "unexpected error with escaped FQDN", nil, removeKnownHost(client, "escaped?client.uyuni.lan", "42"))
```

File: mgrctl/cmd/ssh/remove_known_host_test.go [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+// Test removeKnownHost function.
+func TestRemoveKnownHost(t *testing.T) {
+	client, err := api.Init(connectionDetails)
+	if err != nil {
+		t.FailNow()
+	}
+
+	client.Client = &mocks.MockClient{
+		DoFunc: MockKnownHostDo,
+	}
```
Comment (cbosdo) [inline mgrctl/cmd/ssh/remove_known_host_test.go]:
```suggestion
	client.Client = &mocks.MockClient{
		DoFunc: func (eq *http.Request) (*http.Response, error) {
		  testutils.AssertEquals(t, "Wrong URL called", req.URL.Path, "/rhn/manager/api/admin/ssh/removeKnownHost")
		  // You can use asserts to check the passed parameters directly here
		  // You can return whatever response you want to even simulate errors from the server
		},
	}
```

When combined with a test loop like https://github.com/uyuni-project/uyuni-tools/pull/692/changes#diff-8a8355369c4782d28a8b743cc823d376f45dbb33dfc6f51aa3ee9abccf5b9998 this makes it super easy to add more corner cases.

File: mgrctl/cmd/ssh/remove_known_host_test.go [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+			if err != nil {
+				t.FailNow()
+			}
+
+			client.Client = &mocks.MockClient{
+				DoFunc: func(req *http.Request) (*http.Response, error) {
+					testutils.AssertEquals(t, "Wrong URL called", req.URL.Path, "/rhn/manager/api/admin/ssh/removeKnownHost")
+
+					query := req.URL.Query()
+					testutils.AssertTrue(t, "Both parameters have to be given", query.Has("hostname") && query.Has("port"))
```
Comment (cbosdo) [inline mgrctl/cmd/ssh/remove_known_host_test.go]:
```suggestion
					testutils.AssertEquals(t, "The hostname is not properly passed ", url.QueryEscape(tt.hostname)), query.Get("hostname"))
					testutils.AssertEquals(t, "The port is not properly passed ", tt.port, query.Get("port"))
```

Would probably make more sense so that you are sure the data passed to the server are what you expect.

File: mgrctl/cmd/ssh/remove_known_host_test.go [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+
+					// Check if the port is a number
+					_, err := strconv.Atoi(query["port"][0])
+					if err == nil && !tt.expectedFail {
+						code = 200
+						body = `{"success":true,"result":1}`
+					}
+
+					testutils.AssertEquals(t, "Expected to return with different status code", tt.statusCode, code)
+					return testutils.GetResponse(code, body)
```
Comment (cbosdo) [inline mgrctl/cmd/ssh/remove_known_host_test.go]:
```suggestion
					return testutils.GetResponse(tt.statusCode, tt.body)
```

would give full control on the mocked result: no need to check any other data validity as that would more or less reimplement what is under test.

File: mgrctl/cmd/ssh/remove_known_host_test.go [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+var connectionDetails = &api.ConnectionDetails{User: user, Password: password, Server: server}
+
+// Test removeKnownHost function.
+func TestRemoveKnownHost(t *testing.T) {
+	tests := []struct {
+		name         string
+		hostname     string
+		port         string
+		statusCode   int
+		expectedFail bool
```
Comment (cbosdo) [inline mgrctl/cmd/ssh/remove_known_host_test.go]:
```suggestion
		statusCode   int
		body string
		expectedError string
```

So you can return mocked body and assert the error message

File: mgrctl/cmd/ssh/remove_known_host_test.go [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+	}{
+		{
+			name:       "Test removing normal hostname and port",
+			hostname:   "client.uyuni.lan",
+			port:       "22",
+			statusCode: 200,
```
Comment (cbosdo) [inline mgrctl/cmd/ssh/remove_known_host_test.go]:
```suggestion
			statusCode: 200,
			body: `{"success":true,"result":1}`,
			expectedError: "",
```

File: mgrctl/cmd/ssh/remove_known_host_test.go [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+		},
+		{
+			name:       "Test removing escaped hostname and port",
+			hostname:   "escaped?client.uyuni.lan",
+			port:       "42",
+			statusCode: 200,
```
Comment (cbosdo) [inline mgrctl/cmd/ssh/remove_known_host_test.go]:
```suggestion
			statusCode: 200,
			body: `{"success":true,"result":1}`,
			expectedError: "",
```

File: mgrctl/cmd/ssh/remove_known_host.go [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+
+	success, ok := res.Result.(float64)
+	if !ok {
+		return fmt.Errorf(L("unexpected server response '%v'"), res.Result)
+	}
+
+	if int(success) == 1 {
+		fmt.Println(L("successfully removed host"))
+	} else {
+		fmt.Println(L("unable to remove host"))
```
Comment (cbosdo) [inline mgrctl/cmd/ssh/remove_known_host.go]: Is this reachable? Isn't that covered with the `res.Success` check?

File: mgrctl/cmd/ssh/remove_known_host.go [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+	path := "admin/ssh/removeKnownHost?" + params.Encode()
+
+	var data map[string]interface{}
+	res, err := api.Post[interface{}](client, path, data)
+	if err != nil {
+		return utils.Errorf(err, L("error in query '%s'"), path)
+	}
+
+	if !res.Success {
+		return fmt.Errorf("%s", res.Message)
```
Comment (cbosdo) [inline mgrctl/cmd/ssh/remove_known_host.go]:
```suggestion
		return fmt.Errorf(L("failed to remove the host: %s"), res.Message)
```

File: mgrctl/cmd/ssh/remove_known_host_test.go [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+		},
+		{
+			name:         "Test removing an invalid port value",
+			hostname:     "myotherclient.uyuni.lan",
+			port:         "abc",
+			statusCode:   400,
+			expectedFail: true,
```
Comment (cbosdo) [inline mgrctl/cmd/ssh/remove_known_host_test.go]:
```suggestion
			statusCode:   400,
			body: `{"success":false,"message":"mocked error"}`,
			expectedError: "mocked error",
```

File: mgrctl/cmd/ssh/remove_known_host_test.go [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+
+					testutils.AssertEquals(t, "Expected to return with different status code", tt.statusCode, code)
+					return testutils.GetResponse(code, body)
+				},
+			}
+
+			if err := removeKnownHost(client, tt.hostname, tt.port); err != nil {
+				if !tt.expectedFail {
+					t.Errorf("%s failed: %v", tt.name, err)
+				}
```
Comment (cbosdo) [inline mgrctl/cmd/ssh/remove_known_host_test.go]:
```suggestion
				if !tt.expectedError {
					t.AssertStringContains(t, "Unexpected error message", err.Error(), tt.expectedError)
				}
```

File: mgrctl/cmd/ssh/remove_known_host.go (hunk 14 above)
Comment (marv7000) [reply mgrctl/cmd/ssh/remove_known_host.go]: IIRC the server can return 0 if e.g. there is no known_hosts file

Comment (cbosdo) [reply mgrctl/cmd/ssh/remove_known_host.go]: Then may be this should be more explicit in the error message

File: mgrctl/cmd/ssh/knownhost/remove_test.go [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+			if err != nil {
+				t.FailNow()
+			}
+
+			client.Client = &mocks.MockClient{
+				DoFunc: func(req *http.Request) (*http.Response, error) {
+					testutils.AssertEquals(t, "Wrong URL called", req.URL.Path, "/rhn/manager/api/admin/ssh/removeKnownHost")
+
+					query := req.URL.Query()
+					testutils.AssertEquals(t, "The hostname is not properly passed", tt.hostname, query.Get("hostname"))
```
Comment (cbosdo) [inline mgrctl/cmd/ssh/knownhost/remove_test.go:70]: Isn't the hostname supposed to be URL-encoded at this point?

**Review by cbosdo — APPROVED** [review]
